* Add support for reading and extracting encrypted archives (memory a…

udgover opened the pull request:
Hello, this PR adds support to read and extract encrypted archives. Tests were successful for each libarchive.public reader.

Decrypting such archives is as simple as providing passphrase argument to functions.

dsoprea commented:
That API wasn't added until libarchive 3.1.900 (https://github.com/libarchive/libarchive/commit/6c222e59f461bf61962c7de318f946147f58d29b) and Ubuntu didn't move past 3.1.2 until it 3.2.1-2 was backported into Xenial, and, even then, it's in scope but not the default installed version. So, I updated the distribution, specifically installed the correct version of the library, and added a bunch of version and API verbosity in the CI output for quicker triaging in the future: https://github.com/dsoprea/PyEasyArchive/commit/b4f19a05e91755596ced928f8630ce5f47c7f611

I also had to drop Python 3.3 support (since it's not available for Xenial), add support for Python 3.6 and 3.7, and fix an encoding issue with some follow-up commits. Now everything should be green.

I need you to rebase your change on top of latest upstream.

udgover commented:
Hello,

Thanks for your feedback. I've just rebased my commit.

dsoprea commented on libarchive/adapters/archive_read_add_passphrase.py:
```diff
@@ -0,0 +1,5 @@
+import libarchive.calls.archive_read_add_passphrase
```
This file is not necessary. It's already integrated in adapters/archive_read.py (where higher-level read operations are integrated).

udgover replied:
I've just mimicked your original code for other API call. Let me know how to handle it.

dsoprea commented on libarchive/adapters/archive_read.py:
```diff
@@ -267,14 +275,15 @@ def _set_read_context(archive_res, format_code=None, filter_code=None):
     archive_res = _archive_read_new()
 
     try:
+        r = _archive_read_add_passphrase(archive_res, passphrase)
```
Only call this if not `None`.

udgover replied:
At first, I write a condition to check if None but after doing some tests, providing None to the API call does not raise an error.

kbalk replied:
Couple of comments:

- By saying "only call this if not "None", do you really mean `if passphrase:`?
- 'r' is a success/failure code, but it isn't checked for failure.  But I'm not sure how a failure would be handled here.  
- The call below this (which was not modified) never returns a value, so this 'r' should be omitted:
    `r = _set_read_context(archive_res, format_code, filter_code)`

dsoprea commented on libarchive/adapters/archive_read.py:
```diff
@@ -49,6 +49,14 @@ def _archive_read_support_format_all(archive):
         raise libarchive.exception.ArchiveError(message)
 
+def _archive_read_add_passphrase(archive, passphrase):
+    try:
```
Encode passphrase to bytes first: `passphrase.encode('utf-8')`

kbalk replied:
That will cause Python 3 to error ...

dsoprea commented on libarchive/adapters/archive_read.py:
```diff
@@ -12,7 +12,7 @@
 import libarchive.adapters.archive_read_set_format
 import libarchive.adapters.archive_read_support_format_all
-
+import libarchive.adapters.archive_read_add_passphrase
```
Not necessary. See below.

udgover commented:
I will add the interfaces for adding a passphrase while creating/updating and provide some tests.

kbalk commented:
The C libarchive permits multiple passphrases - could you add support for more than one passphrase, i.e., instead of passing a single passphrase, allow for a list of passphrases?

dsoprea commented:
@kbalk Interesting idea.
@udgover Bump.

udgover commented:
Hello,

Long time no see, sorry... I've finally got some time to implement write support with archive_write_set_passphrase. It works like a charm and I just need to write tests and I'll submit the PR.

udgover commented:
I'm about to clean the mess in this PR, add all necessary tests and provide examples.

Currently, I've also added support to provide several passphrases at once. I've replaced passphrase with passphrases in archive_read._enumerator function which then iterate on the provided list of passphrases and call _archive_read_add_passphrase for each element.

I wonder what is the best implementation between completely dropping the possibility to provide only one passphrase as a string and / or providing a list of passphrases.

Both solutions could also coexist but the prototype of the function would be weird:

`def _enumerator(opener, entry_cls, passphrase=None, passphrases=None, format_code=None, filter_code=None): 
`
How to handle the case where both are provided? Take into account both? Raise an error?

Thanks for your feedback.

dsoprea commented:
You could provide a single parameter that accepts both a scalar or a list,
but I'm not a fan of that. You could provide a single parameter that
accepts an object with a "get_passphrases" method, but that's
over-prohramming the solution. Better to always just require a list. That
works for 0, 1, or N passphrases.

Is the use-case to attempt to match one of the passphrases to the archive?
The first match wins?

On Thu, May 28, 2020, 09:59 udgover <notifications@github.com> wrote:

> I'm about to clean the mess in this PR, add all necessary tests and
> provide examples.
>
> Currently, I've also added support to provide several passphrases at once.
> I've replaced passphrase with passphrases in archive_read._enumerator
> function which then iterate on the provided list of passphrases and call
> _archive_read_add_passphrase for each element.
>
> I wonder what is the best implementation between completely dropping the
> possibility to provide only one passphrase as a string and / or providing a
> list of passphrases.
>
> Both solutions could also coexist but the prototype of the function would
> be weird:
>
> def _enumerator(opener, entry_cls, passphrase=None, passphrases=None,
> format_code=None, filter_code=None):
> How to handle the case where both are provided? Take into account both?
> Raise an error?
>
> Thanks for your feedback.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dsoprea/PyEasyArchive/pull/34#issuecomment-635367667>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOW2VSAHK36DS5KG7KRH3RTZU37ANCNFSM4HHE4CQQ>
> .
>

udgover commented:
I'm not a fan with the single parameter for which I have to check type, I could implement an object Passphrases with setter and getter but like you, I think it's over-programming. The list solution looks like to be the best one and I've already done this way for testing.

Concerning the list for encrypted archives, the first match wins.

> You could provide a single parameter that accepts both a scalar or a list, but I'm not a fan of that. You could provide a single parameter that accepts an object with a "get_passphrases" method, but that's over-prohramming the solution. Better to always just require a list. That works for 0, 1, or N passphrases. Is the use-case to attempt to match one of the passphrases to the archive? The first match wins?

udgover commented:
Closing this old messy PR to create another one based on latest commits and taking into account comments of this PR.
